specification/sbi_atee.adoc

```diff
@@ -10,8 +39,1238 @@ shown in the figure above, the TEE Runtime (operating in VS-mode) supports
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code            | Description
+| SBI_ERR_SUCCESS       | The TVM exited, and sbiret.value contains 0 if the
+                        | interrupt or exception is resumable. The host can
+                        | examine `scause` to determine details.
+| SBI_ERR_INVALID_PARAM | `tvm_guest_id` or `tvm_vcpu_id` were invalid, or the
+                          TVM wasn't in `TVM_RUNNABLE` state.
+| SBI_ERR_FAILED        | The operation failed for unknown reasons.
+|===
+
+The TSM updates the `scause` field in the `tvm_vcpu_supervisor_csrs` region in the
+shared-memory for the VCPU that was previously configured by the host. The host should
+use the `scause` field to determine whether the exit was caused by an interrupt or exception,
+and then use the additional information to in the shared-memory region to determine further
+course of action (if sbiret.value is 0).
+
+The TSM sets the most significant bit in `scause` to indicate that that the exit was caused
+by an interrupt, and if this bit is clear, the implication is that the the exit was caused
+by an exception. The remaining bits specific information about the interrupt or exception,
+and the specific reason can be determined using the enumeration detailed below.
+
+[source, C]
+-------
+enum tvm_interrupt_exit {
+    /* The TVM encountered a user-mode software interrupt (informational only; non-resumable). */
+    USER_SOFT = 0,
+    /* The TVM encountered a supervor-mode software interrupt (informational only; non-resumable). */
+    SUPERVISOR_SOFT = 1,
+    /* The TVM encountered a host-mode software interrupt (informational only; non-resumable). */
+    VIRTUAL_SUPERVISOR_SOFT = 2,
+    /* The TVM encountered a machine-mode software interrupt (informational only; non-resumable). */
+    MACHINE_SOFT = 3,
+    /* The TVM encountered a user-mode timer interrupt (informational only; non-resumable). */
+    USER_TIMER = 4,
+    /* The TVM encountered a supervisor-mode timer interrupt (informational only; non-resumable). */
+    SUPERVISOR_TIMER = 5,
+    /* The TVM encountered a host-mode timer interrupt (informational only; non-resumable). */
+    VIRTUAL_SUPERVISOR_TIMER = 6,
+    /* The TVM encountered a machine-mode timer interrupt (informational only; non-resumable). */
+    MACHINE_TIMER = 7,
+    /* The TVM encountered a user-mode interrupt (informational only; non-resumable). */
+    USER_EXTERNAL = 8,
+    /* The TVM encountered a supervisor-mode interrupt (informational only; non-resumable). */
+    SUPERVISOR_EXTERNAL = 9,
+    /* The TVM encountered a host-mode interrupt (informational only; non-resumable). */
+    VIRTUAL_SUPERVISOR_EXTERNAL = 10,
+    /* The TVM encountered a machine-mode interrupt (informational only; non-resumable). */
+    MACHINE_EXTERNAL = 11,
+    /* The TVM encountered a supervisor-mode interrupt (informational only; non-resumable). */
+    SUPERVISOR_GUEST_EXTERNAl = 12,
```

Comment:
general Q: whats the difference between supervisor guest external interrupt and VS_External interrupt?

Reply:
Good question -- this is from the Salus code -- @abrestic-rivos might know for sure.

Reply:
Virtual supervisor external interrupt: an external interrupt directed at VS level.
Supervisor guest external interrupt: an HS-level interrupt indicating that are pending VS external interrupts.

The second is necessary because only one guest interrupt file may be active at a time, and the host still needs to know if an interrupt is received for one of the inactive interrupt files. See the privileged and AIA specs for more details.

BTW -- these exception/interrupt causes are simply the ones defined in the privileged spec. Not sure it's worth repeating them here. We should just refer to the privileged spec instead.